File: tests/python/binary_trees/find_nodes_distance_k/test_find_nodes_distance_k.py
from problems.binary_trees.find_nodes_distance_k.python.find_nodes_distance_k import (
    BinaryTree, findNodesDistanceK)
from tests.python.utility import load_test_cases


def build_bt(nodes: list[dict], root_id: str) -> BinaryTree:
    id_to_node = {}

    for node in nodes:
        id_to_node[node["id"]] = BinaryTree(node["value"])

    for node in nodes:
        bt_node = id_to_node[node["id"]]
        bt_node.left = id_to_node.get(node["left"])
        bt_node.right = id_to_node.get(node["right"])

    return id_to_node[root_id]


def test_find_nodes_distance_k() -> None:
    test_cases = load_test_cases("binary_trees/find_nodes_distance_k.json")

    for idx, test in enumerate(test_cases):
        tree = test["tree"]
        target = test["target"]
        k = test["k"]
        root = build_bt(tree["nodes"], tree["root"])
        expected = test["expected"]

        result = findNodesDistanceK(root, target, k)

        assert (
            result.sort() == expected.sort()
        ), f"Test case {idx} failed: expected {expected}, got {result}"

File: tests/cpp/binary_trees/test_find_nodes_distance_k.cpp
#include "../../../problems/binary_trees/find_nodes_distance_k/cpp/find_nodes_distance_k.h"
#include "../jsontestbase.h"
#include "./utility.h"

#include <algorithm>
#include <vector>

class FindNodesDistanceKTest : public JsonTestBase {
public:
  FindNodesDistanceKTest() {
    json_file_path = std::string(TEST_CASES_DIR) +
                     "/binary_trees/find_nodes_distance_k.json";
  }
};

TEST_F(FindNodesDistanceKTest, TestCases) {
  ASSERT_FALSE(test_cases.empty()) << "Test cases are empty";

  for (const auto &test : test_cases) {
    auto treeJson = test["tree"];
    auto target = test["target"].get<int>();
    auto k = test["k"].get<int>();
    auto expected = test["expected"].get<std::vector<int>>();

    auto root = buildBT(treeJson);
    auto result = findNodesDistanceK(root, target, k);

    std::sort(result.begin(), result.end());
    std::sort(expected.begin(), expected.end());

    EXPECT_EQ(result, expected) << "Failed for test case: " << test.dump(2);

    deleteBT(root);
  }
}
